### PR TITLE
Added ASP.NET Core Middleware

### PR DESCRIPTION
**Pull request by randyridge**
Added some simple middleware for GraphQL and GraphiQL endpoints.  Need to add unit tests.

GraphQL still needs some work with dependency injection for schema.

GraphiQL is stuck to 0.7.5 until GraphQL support is on par with 0.7.0.

usage:

```
            app.UseGraphQL(new GraphQLOptions {
                GraphQLPath = "/graphql",
                Schema = new Schema()
            });
            app.UseGraphiQL(new GraphiQLOptions {
                GraphiQLPath = "/graphiql",
                GraphQLPath = "/graphql"
            });
```

### CODE REVIEW

**Comment by joemcbride** [comment]
I was able to fix the bug that was causing the issue with GraphiQL.  The dotnet core project has been updated to reflect that, and I bumped the versions of graphql-js and GraphiQL as well.

I like this idea.  Were you thinking these would be code samples, ship with the core lib, or ship as a new nuget?

**Comment by randyridge** [comment]
i just set them up to do some quick prototyping.  I think they'd work fine as nuget packages to get folks up and running quickly and for examples.

**Comment by wesamco** [comment]
in a MVC application, you might want to short-circuit the middleware pipeline after you return the response
by not calling 
`await next(context).ConfigureAwait(true);`

File: src/GraphQL.AspNetCore.GraphQL/GraphQLMiddleware.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            return await new DocumentExecuter().ExecuteAsync(schema, null, graphqlRequest.Query, graphqlRequest.OperationName, graphqlRequest.Variables.ToInputs()).ConfigureAwait(true);
+        }
+
+        private bool ShouldRespondToRequest(HttpRequest request) {
+            return string.Equals(request.Method, "POST", StringComparison.OrdinalIgnoreCase) && request.Path.Equals(graphqlPath);
+        }
+
+        private static Task WriteResponseAsync(HttpResponse response, ExecutionResult executionResult) {
+            response.ContentType = "application/json";
+            response.StatusCode = executionResult.Errors?.Count == 0 ? 200 : 400;
```
Comment (mirhagk) [inline src/GraphQL.AspNetCore.GraphQL/GraphQLMiddleware.cs]: Wouldn't this make having no errors (errors being null) default to showing error 400? That's what I'm getting when I run this locally. The result is working perfectly fine however.

Should switch this to `response.StatusCode = executionResult.Errors?.Count > 0 ? 400 : 200;`

Comment (randyridge) [inline src/GraphQL.AspNetCore.GraphQL/GraphQLMiddleware.cs]: looks like it, i'll hopefully get back to graphql stuff next week

**Comment by randyridge** [comment]
i'm going to get these up-to-date with tests next week.

File: src/GraphQL.AspNetCore.GraphQL/GraphQLMiddleware.cs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        ///     Throws <see cref="ArgumentNullException" /> if <paramref name="context" />.
+        /// </exception>
+        public async Task Invoke(HttpContext context) {
+            if(context == null) {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if(ShouldRespondToRequest(context.Request)) {
+                var executionResult = await ExecuteAsync(context.Request).ConfigureAwait(true);
+                await WriteResponseAsync(context.Response, executionResult).ConfigureAwait(true);
```
Comment (Kayomani) [inline src/GraphQL.AspNetCore.GraphQL/GraphQLMiddleware.cs]: You shouldn't call next if you have responded here.  Whats with the .ConfigureAwait(true) all over the place? There shouldn't be any need to keep the original context.

This and the errors check previously discussed are the only two bugs I have I found with this PR. GraphQL.AspNetCore.GraphiQL.index.html seems to be missing too but I didn't test the explorer middleware.

**Comment by joemcbride** [comment]
I'm going to close this now as some middleware has been made available as a Nuget.

**Pull request closed without merging**